### ISSUE TITLE
Use a task instead of a test for binary breakage checking

### ISSUE THREAD

**Issue by wolfs** [gradle/gradle#2378]
Instead of using a test as introduced by #2365, we could also use a dedicated task, similar as provided by @melix's [japicmp-gradle-plugin](https://github.com/melix/japicmp-gradle-plugin) or extend that plugin to provide the necessary functionality.
The advantage would be that we could make this single task cacheable and have a report including all the API changes.
The disadvantages would include that it is more heavy-weight:
- We would need a method to download a previous Gradle version and cache this download. This is already provided by our integration test infrastructure via [DownloadableGradleDistribution](https://github.com/gradle/gradle/blob/ea100084506f169ddc4662087a200bfa275dca81/subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/executer/DownloadableGradleDistribution.groovy).
- We would need to make it possible to provide two different classpaths for the comparison to the plugin, as we do now for the test [here](https://github.com/gradle/gradle/blob/0a55f6c19995109f31a2d8f76c7e6e4db5e74309/subprojects/distributions/src/integTest/groovy/org/gradle/BinaryBreakageIntegrationTest.groovy#L118-L120). The reason is that we also want to detect breakages caused by external APIs bleeding into our public API.
- We would need to provide a way to accept changes by using [custom failure conditions](https://github.com/melix/japicmp-gradle-plugin#custom-build-failure-conditions). This seems to be possible but slightly more involved then the [straightforward solution we use now](https://github.com/gradle/gradle/blob/0a55f6c19995109f31a2d8f76c7e6e4db5e74309/subprojects/distributions/src/integTest/groovy/org/gradle/BinaryBreakageIntegrationTest.groovy#L64-L67).

**Comment by wolfs** [comment]
A comment on the report provided: I don't find a large report with non-actionable items very useful. Even if we would provide this for each release, as a plugin author I would not look through the changes but try out if my plugin works with the new release. I have been checking the JAPI-CMP report for past releases and found it also very hard to track down what happened - even though it is better readable than the JDiff report because it contains some indication what we could have broken.

**Comment by melix** [comment]
Maybe it changed, but when I implemented the plugin, the check for "binary compatible" wasn't as simple as checking the `clazz.isBinaryCompatible()` flag: this would mean that the _class_ itself is binary compatible (same superclass, same interfaces), but not that its methods were. Anyway, I think there's value in a report, and that there's value in having a task.

Downloading a previous version is a no-brainer, that's why we have dependency resolution. Providing two different classpath shouldn't be a problem either (I think).

I'll check what I can do with the plugin, but basically the report is completely customizable. If you want to only report breaking changes, let it go. If you want to show everything, you can. And if you want to show that this method is to be replaced by another, you can also do it.

**Comment by melix** [comment]
FTR I already made some good progress on updating the plugin: https://github.com/melix/japicmp-gradle-plugin/tree/cc/update-to-0.10.0

The API of JApiCmp changed a bit and is smarter wrt binary changes. The branch can already use 2 different classpath, and I've simplified it to always use `FileCollection` as inputs. If you set it to a `Configuration`, then it will compare the first level dependencies. HTML report generation works (using the default template, which is a bit hard to read).

**Comment by wolfs** [comment]
@melix Great progress.

One question regarding your earlier comment:
> Downloading a previous version is a no-brainer, that's why we have dependency resolution. 

Could you please explain how we can download the previous Gradle version with dependency resolution? I am not aware we publish it to a Maven (or Ivy) repository. Even then, we would need to unpack it and obtain the internal (`gradle-*`) and the external dependencies. Caching unpacking of a Gradle distribution is done by our test infrastructure. I do not know about something which provides this functionality out of the box without using the test infrastructure.

**Comment by bamboo** [comment]
> Could you please explain how we can download the previous Gradle version with dependency resolution?

There seems to be some synergy between this task and that idea of publishing stable Gradle API jars to allow plugin authors to target previous Gradle releases while still building their plugins with the very latest Gradle.

**Comment by oehme** [comment]
A major benefit of a task over a test is reusability. Of course, the way the current test is structured it is not very interesting for reuse. But if we generalised this a bit, it could turn into a JapiCmp plugin that our users could benefit from.

That being said, there's no inherent problem with this being a test for now.

**Comment by melix** [comment]
I have made significant progress on the new JApiCmp plugin. Here's an example report that compares `gradle-core-3.5.jar` with `gradle-core-4.0.jar`:

![selection_050](https://user-images.githubusercontent.com/316357/27922336-fb4cc0a4-627a-11e7-9517-831a727e3185.png)

This is fully configurable. The configuration I used for this test is very simple (it will need tweaking, but it's fairly easy):

```groovy
task japicmp(type: JapicmpTask) {
    oldClasspath = files('/home/cchampeau/.sdkman/gradle/3.5/lib/gradle-core-3.5.jar')
    newClasspath = files('/home/cchampeau/.sdkman/gradle/4.0/lib/gradle-core-4.0.jar')
    onlyModified = true
    failOnModification = true
    txtOutputFile = file("$buildDir/reports/japi.txt")
    htmlOutputFile = file("$buildDir/reports/japi.html")
    ignoreMissingClasses = true
    richReport {
        title = 'Binary compatibility report for japicmp'
        def acceptedViolations = [
                'Implemented interface org.gradle.api.attributes.HasConfigurableAttributes': '@Incubating interface has been removed'
        ]
        addRule { JApiCompatibility member ->
            if (!member.binaryCompatible) {
                def acceptation = acceptedViolations[Violation.describe(member)]
                if (acceptation) {
                    Violation.accept(member, acceptation)
                } else {
                    Violation.notBinaryCompatible(member)
                }
            }
        }
        addRule(JApiChangeStatus.NEW) { member ->
            if (member instanceof JApiMethod) {
                if (!member.annotations.find { it.fullyQualifiedName == 'org.gradle.api.Incubating' }) {
                    if (!member.jApiClass.annotations.find { it.fullyQualifiedName == 'org.gradle.api.Incubating'}) {
                        Violation.error(member, "New method is not annotated with @Incubating")
                    }
                }
            }
        }
        includedClasses = [
                'org[.]gradle[.].[^.]+',
                'org[.]gradle[.]api[.].*',
                'org[.]gradle[.]authentication[.].*',
                'org[.]gradle[.]buildinit[.].*',
                'org[.]gradle[.]caching[.].*',
                'org[.]gradle[.]external[.]javadoc[.].*',
                'org[.]gradle[.]ide[.].*',
                'org[.]gradle[.]includedbuild[.].*',
                'org[.]gradle[.]ivy[.].*',
                'org[.]gradle[.]jvm[.].*',
                'org[.]gradle[.]language[.].*',
                'org[.]gradle[.]maven[.].*',
                'org[.]gradle[.]nativeplatform[.].*',
                'org[.]gradle[.]normalization[.].*',
                'org[.]gradle[.]platform[.].*',
                'org[.]gradle[.]play[.].*',
                'org[.]gradle[.]plugin[.]devel[.].*',
                'org[.]gradle[.]plugin[.]repository[.]*',
                'org[.]gradle[.]plugin[.]use[.]*',
                'org[.]gradle[.]plugin[.]management[.]*',
                'org[.]gradle[.]plugins[.].*',
                'org[.]gradle[.]process[.].*',
                'org[.]gradle[.]testfixtures[.].*',
                'org[.]gradle[.]testing[.]jacoco[.].*',
                'org[.]gradle[.]tooling[.].*',
                'org[.]gradle[.]model[.].*',
                'org[.]gradle[.]testkit[.].*',
                'org[.]gradle[.]testing[.].*']
        excludedClasses = [
                '[.]internal[.]'
        ]
    }
}
```

**Comment by melix** [comment]
One step further. I used:

1. a custom Ivy repository to download the distribution
2. a dependency transform, which extracts a Gradle module jar from the distribution
3. a configuration per jar, as well as a japicmp task per jar

This allowed me to generate the attached reports.
[reports.zip](https://github.com/gradle/gradle/files/1128955/reports.zip)

For the curious, here's how I did this: https://gist.github.com/anonymous/a462a1e805e983240ab0d1ad2efedc2c

Isn't it nice? :)
